rfc/rfc-64/rfc-64.md

```diff
@@ -0,0 +1,509 @@
+interface TimelineView {
+    // Looks up latest commit instant, provide (temporal) instant T;
+    // This method is required to resolve commit-instants by the timestamp
+    HoodieInstant findLatestCompletedActionAt(Instant instant);
+    
+    // ...
+}
+```
+
+##### File Listing
+
+To provide for *listing* capability described above we will be introducing following facades:
+ 
+ - **RecordIndex**: to be able to resolve records against the file-groups where such records
+are persisted (*required for writing*)
+ - **FileIndex**: to be able to list file(-slices) constituting *partition snapshot*, satisfying 
+pushed-down data-filters (validated t/h Partition Pruning, Files' Columns Stats filtering, etc)
+
+```java
+class RecordIndex {
+    // Tags provided record-keys (PKs) within corresponding locations within the dataset;
+    // in case record-key is not currently present in the dataset, no location will be provided
+    HoodiePairData<HoodieKey, Option<HoodieRecordLocation>> tag(HoodieData<HoodieKey> records);
+}
+
+class FileIndex {
+    // Lists files visible/reachable at the instant T
+    // Equivalent to `listFilesBetween(null, instant, filters)`
+    List<PartitionSnapshot> listFilesAt(HoodieInstant instant, Predicate[] filters);
+    // Lists files added visible/reachable at the instant `to`, that were
+    // added no earlier than at the instant `from`
+    List<PartitionIncrementalSnapshot> listFilesBetween(HoodieInstant from, HoodieInstant to, Predicate[] filters);
+    
+    // TODO add CDC-capable API
+}
+```
+
+##### Reading
+
+Reading capability will be provided by `FileSliceReader` allowing for:
+
+ - Data filters push-down
+ - Schema projections (for columnar-based formats)
+ - Schema evolution handling
+ - Producing records in different modes (snapshot/incremental/CDC)
+
+```java
+class FileSliceReader {
+    // Projects output of this reader as a projection of the provided schema
+    // NOTE: Provided schema could be an evolved schema
+    FileSliceReader project(InternalSchema schema);
+    // Pushes down filters to low-level file-format readers (if supported)
+    FileSliceReader pushDownFilters(Predicate[] filters);
+    // Specifies reading mode for this reader
+    FileSliceReader readingMode(ReadingMode mode);
+    // Produces an iterable sequence of records from the particular file-slice
+    Iterator<HoodieRecord> open(HoodieFileSlice fileSlice);
+}
+```
+
+To see how `FileSliceReader` will be used, let's take example of Presto:
+
+1. `HoodieMergedLogRecordScanner#getRecords` API will be returning
+opaque `HoodieRecord`s as returned by `FileSliceReader#open`.
+2. `HoodieRealtimeRecordReader` will implement Presto's `RecordReader` APIs using the
+aforementioned log scanner API. Then, in Presto we can create `RecordCursor`
+wrapping around `RecordReader`. The key advantage here is that
+the `HoodieRealtimeRecordReader` won't have to spend cycles
+converting `ArrayWritable` to `Avro` and vice-versa as it does today.
```

Comment:
Currently for hive/presto, we use MapredParquetInputFormat to read the parquet file and return ArrayWritable records.
Do we want to implement our own Parquet reading logic and return avro type directly to avoid spend cycles
converting `ArrayWritable` to `Avro`？

Reply:
@xiarixiaoyao Good question! We don't intend to implement our own Parquet reading logic. The work here will build on top of [RFC-46](https://github.com/apache/hudi/blob/master/rfc/rfc-46/rfc-46.md). So, in this case, we will have an implementation of `HoodieRecord` and `HoodieRecordMerger`( as proposed in RFC-46) for Hive `ArrayWritable`.

Reply:
Will custom HoodieRecordMerger implementations be supported and if so how will the reader load those custom implementations that are not part of open source?

Reply:
@the-other-tim-brown this is covered in [RFC-46](https://github.com/apache/hudi/blob/master/rfc/rfc-46/rfc-46.md): If you want to implement custom merging logic, then

 - When creating your table you specify unique MERGER_STRATEGY uuid (identifying your merging strategy/semantic)
 - During writing you specify MERGER_IMPLS, which are actual engine-specific RecordMerger impls, that realize particular merging strategy (set as MERGER_STRATEGY)
 - When writing Hudi will take up MERGER_IMPLS, filter the ones based on the MERGER_STRATEGY uuid, and use them for merging (based on which engine it's running on)